scripts/symbolize.py

```diff
@@ -162,6 +166,9 @@ def elf_load_addr(self, addr):
                     if elf_idx is not None:
                         return self._elfs[int(elf_idx)][1]
             return None
+        elif self.load_addr:
```

Comment:
> How does this work with shared libraries? I suspect it doesn't ;-)

Correct. I haven't tested this with shared libraries.

> One way to fix that could be to dump the mapped regions at the beginning of the trace buffer (`user_ta_dump_state()`, except that this one sends its output to the console, not to a buffer...).

Hmm, it seems that I need to replicate `user_ta_dump_state()` like `user_ta_dump_state_buffer()` that would dump output in ftrace buffer rather that console. So I could reuse the existing region parsing framework in `symbolize.py`.

Reply:
I am facing a linker error in case both `CFG_ULIBS_SHARED` and `CFG_ULIBS_MCOUNT (earlier CFG_ULIBS_GPROF)` as:
```
out/ta.lds:19: unresolvable symbol `__utee_mcount' referenced in expression
```
It seems like `-pg` profiling option doesn't support profiling of shared libraries. Maybe the reason being who will provide `mcount` api definition for various shared libraries linked at runtime. Also this seems to be the reason for GPROF not able to profile shared libraries, see: https://stackoverflow.com/questions/1838989/gprof-how-to-generate-call-graph-for-functions-in-shared-library-that-is-linke

It seems that we won't be able to support ftrace for shared libraries too. So I think we should add a config error check that `CFG_ULIBS_MCOUNT=y` can't coexist with `CFG_ULIBS_SHARED=y`.

Reply:
I don't think there is a fundamental problem here. It's probably just the hack that we are using in the linker script that isn't valid when `__utee_mcount` is in  shared library.

There must be some way around that. Best thing would be to drop this whole "pre-allocate buffer space in .bss" stuff. We will be able do that when we have an API (syscall) to request TA memory from the TEE core. As far as I can tell, this is coming as a followup to Jens' series on shared pages (which introduces nice improvements to the memory mapping code). But in the mean time, I'm not sure how to proceed. Disabling profiling when `CFG_ULIBS_SHARED=y` may be the most reasonable option, but we should try to find a workaround first. I have no idea now but will think about it.

Reply:
https://github.com/OP-TEE/optee_os/pull/2943

Reply:
> I don't think there is a fundamental problem here. It's probably just the hack that we are using in the linker script that isn't valid when __utee_mcount is in shared library.

It seems to be a fundamental problem to me. If `__utee_mcount` be part of TA itself rather than shared library than also I don't think APIs in shared library would be able to dynamically link to `__utee_mcount` in TA. AFAIK, it is the application which can dynamically link to shared library APIs and reverse is not true.

> Best thing would be to drop this whole "pre-allocate buffer space in .bss" stuff. We will be able do that when we have an API (syscall) to request TA memory from the TEE core.

I don't think this issue is related to memory allocation.

> Disabling profiling when CFG_ULIBS_SHARED=y may be the most reasonable option

We can still profile TA (via `CFLAGS_ta_arm64=-pg`) but not the shared library code (`CFG_ULIBS_MCOUNT != y`).

> I'll post a PR to prevent `CFG_ULIBS_SHARED` and `CFG_ULIBS_GPROF` from being enabled simultaneously.

Thanks.

Reply:
> AFAIK, it is the application which can dynamically link to shared library APIs and reverse is not true.

Wrong ;-) A shared lib may have undefined symbols. The runtime linker (ELF loader) is responsible for resolving them. See http://refspecs.linuxbase.org/elf/gabi4+/ch5.dynamic.html#shobj_dependencies.
OP-TEE scans the binaries starting from the main every time it processes relocations for a binary file, see https://github.com/OP-TEE/optee_os/blob/master/core/arch/arm/kernel/user_ta.c#L642. So if a library uses a symbol that is defined in the TA main executable, it will resolve properly.

> I don't think this issue is related to memory allocation.

Not really, what I meant was, the "hack" in the linker script that causes the error was largely introduced to detect if -pg is used in the TA and if so, reserve some space in `.bss`. By rewriting this altogether, we can avoid the problem.

Reply:
Ah ok, its good to know this dynamic linker feature. BTW, thanks for the references.

> By rewriting this altogether, we can avoid the problem.

I suppose then we need to separate out the profiling part from libutee and make it a separate library which is always statically linked with TA when profiling is enabled.

And for `.bss` section hack, I am not sure how profiling of those library apis would work which will be responsible for memory allocation. It seems to be chicken-egg situation here.

Reply:
> I suppose then we need to separate out the profiling part from libutee and make it a separate library which is always statically linked with TA when profiling is enabled.

Hopefully not needed. The mcount API can hopefully be in a shared library, and it should not matter, it is only its use in a linker script with PROVIDE(MCOUNT_SYM = __utee_mcount) that is problematic I think. Hence my proposal to remove this "hack", which sole purpose is to allocate and use the profiling buffer only if mcount is called by the TA.

> And for `.bss` section hack, I am not sure how profiling of those library apis would work which will be responsible for memory allocation. It seems to be chicken-egg situation here.

Buffer would be allocated by the TA's initialization function. We would need to find a way to detect if mcount is actually used anywhere in the TA. Some help from the loader might be needed.

Reply:
> The mcount API can hopefully be in a shared library, and it should not matter, it is only its use in a linker script with PROVIDE(MCOUNT_SYM = __utee_mcount) that is problematic I think.

Hmm, understood. This can't detect `__utee_mcount` from shared library.

Can't we directly define  `_mcount` (for aarch64) or `__gnu_mcount_nc` (for arm) rather than providing them via `__utee_mcount`? I have tried to do so and I didn't find any issue with shared library. Also I could get rid of hack for `__utee_mcount` in `lib/libutee/arch/arm/user_ta_entry.c +19`. It seems like compiler and linker treats `_mcount` (for aarch64) or `__gnu_mcount_nc` (for arm) as special symbols and `__utee_mcount` as normal symbol (could be discarded if not used).

Reply:
See: https://github.com/OP-TEE/optee_os/pull/2945